### Flight Delays

In [1]:
import pandas as pd
import numpy as np

In [2]:
df = pd.read_csv('flights.csv')

C:\Users\User1\Anaconda3\lib\site-packages\IPython\core\interactiveshell.py:3058: DtypeWarning: Columns (7,8) have mixed types. Specify dtype option on import or set low_memory=False.
  interactivity=interactivity, compiler=compiler, result=result)


In [3]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 5819079 entries, 0 to 5819078
Data columns (total 31 columns):
YEAR                   int64
MONTH                  int64
DAY                    int64
DAY_OF_WEEK            int64
AIRLINE                object
FLIGHT_NUMBER          int64
TAIL_NUMBER            object
ORIGIN_AIRPORT         object
DESTINATION_AIRPORT    object
SCHEDULED_DEPARTURE    int64
DEPARTURE_TIME         float64
DEPARTURE_DELAY        float64
TAXI_OUT               float64
WHEELS_OFF             float64
SCHEDULED_TIME         float64
ELAPSED_TIME           float64
AIR_TIME               float64
DISTANCE               int64
WHEELS_ON              float64
TAXI_IN                float64
SCHEDULED_ARRIVAL      int64
ARRIVAL_TIME           float64
ARRIVAL_DELAY          float64
DIVERTED               int64
CANCELLED              int64
CANCELLATION_REASON    object
AIR_SYSTEM_DELAY       float64
SECURITY_DELAY         float64
AIRLINE_DELAY          float64
LATE_AIRCRAFT

In [4]:
df.ORIGIN_AIRPORT= df.ORIGIN_AIRPORT.astype(str)

In [5]:
df.DESTINATION_AIRPORT= df.DESTINATION_AIRPORT.astype(str)

In [6]:
df = df[df['ORIGIN_AIRPORT'].apply(lambda x: len(str(x)) <= 3)]
#df = df[df['DESTINATION_AIRPORT'].apply(lambda x: len(str(x)) <= 3)]

In [7]:
origin_unique = df['ORIGIN_AIRPORT'].unique()
destination_unique = df['DESTINATION_AIRPORT'].unique()

In [8]:
airlines_all = df['AIRLINE'].unique()
airlines_all

array(['AS', 'AA', 'US', 'DL', 'NK', 'UA', 'HA', 'B6', 'OO', 'EV', 'MQ',
       'F9', 'WN', 'VX'], dtype=object)

In [10]:
df

YEAR  MONTH  DAY  DAY_OF_WEEK AIRLINE  FLIGHT_NUMBER TAIL_NUMBER  \
0        2015      1    1            4      AS             98      N407AS   
1        2015      1    1            4      AA           2336      N3KUAA   
2        2015      1    1            4      US            840      N171US   
3        2015      1    1            4      AA            258      N3HYAA   
4        2015      1    1            4      AS            135      N527AS   
...       ...    ...  ...          ...     ...            ...         ...   
5819074  2015     12   31            4      B6            688      N657JB   
5819075  2015     12   31            4      B6            745      N828JB   
5819076  2015     12   31            4      B6           1503      N913JB   
5819077  2015     12   31            4      B6            333      N527JB   
5819078  2015     12   31            4      B6            839      N534JB   

        ORIGIN_AIRPORT DESTINATION_AIRPORT  SCHEDULED_DEPARTURE  ...  \
0                  ANC                 SEA                    5  ...   
1                  LAX                 PBI                   10  ...   
2                  SFO                 CLT                   20  ...   
3                  LAX                 MIA                   20  ...   
4                  SEA                 ANC                   25  ...   
...                ...                 ...                  ...  ...   
5819074            LAX                 BOS                 2359  ...   
5819075            JFK                 PSE                 2359  ...   
5819076            JFK                 SJU                 2359  ...   
5819077            MCO                 SJU                 2359  ...   
5819078            JFK                 BQN                 2359  ...   

         ARRIVAL_TIME  ARRIVAL_DELAY  DIVERTED  CANCELLED  \
0               408.0          -22.0         0          0   
1               741.0           -9.0         0          0   
2               811.0            5.0         0          0   
3               756.0           -9.0         0          0   
4               259.0          -21.0         0          0   
...               ...            ...       ...        ...   
5819074         753.0          -26.0         0          0   
5819075         430.0          -16.0         0          0   
5819076         432.0           -8.0         0          0   
5819077         330.0          -10.0         0          0   
5819078         442.0            2.0         0          0   

         CANCELLATION_REASON  AIR_SYSTEM_DELAY  SECURITY_DELAY  AIRLINE_DELAY  \
0                        NaN               NaN             NaN            NaN   
1                        NaN               NaN             NaN            NaN   
2                        NaN               NaN             NaN            NaN   
3                        NaN               NaN             NaN            NaN   
4                        NaN               NaN             NaN            NaN   
...                      ...               ...             ...            ...   
5819074                  NaN               NaN             NaN            NaN   
5819075                  NaN               NaN             NaN            NaN   
5819076                  NaN               NaN             NaN            NaN   
5819077                  NaN               NaN             NaN            NaN   
5819078                  NaN               NaN             NaN            NaN   

         LATE_AIRCRAFT_DELAY  WEATHER_DELAY  
0                        NaN            NaN  
1                        NaN            NaN  
2                        NaN            NaN  
3                        NaN            NaN  
4                        NaN            NaN  
...                      ...            ...  
5819074                  NaN            NaN  
5819075                  NaN            NaN  
5819076                  NaN            NaN  
5819077                  NaN            NaN  
5819078

### Merge with the airline and airport data csvs

In [11]:
airlines = pd.read_csv("airlines.csv")

In [12]:
airports = pd.read_csv("airports.csv")

In [13]:
airlines
# IATA_CODE = Airline in Main CSV

IATA_CODE                       AIRLINE
0         UA         United Air Lines Inc.
1         AA        American Airlines Inc.
2         US               US Airways Inc.
3         F9        Frontier Airlines Inc.
4         B6               JetBlue Airways
5         OO         Skywest Airlines Inc.
6         AS          Alaska Airlines Inc.
7         NK              Spirit Air Lines
8         WN        Southwest Airlines Co.
9         DL          Delta Air Lines Inc.
10        EV   Atlantic Southeast Airlines
11        HA        Hawaiian Airlines Inc.
12        MQ  American Eagle Airlines Inc.
13        VX                Virgin America

In [14]:
airlines = airlines.rename(columns={'IATA_CODE': 'AIRLINE', 'AIRLINE': 'airline_full_name'})

In [15]:
airlines

AIRLINE             airline_full_name
0       UA         United Air Lines Inc.
1       AA        American Airlines Inc.
2       US               US Airways Inc.
3       F9        Frontier Airlines Inc.
4       B6               JetBlue Airways
5       OO         Skywest Airlines Inc.
6       AS          Alaska Airlines Inc.
7       NK              Spirit Air Lines
8       WN        Southwest Airlines Co.
9       DL          Delta Air Lines Inc.
10      EV   Atlantic Southeast Airlines
11      HA        Hawaiian Airlines Inc.
12      MQ  American Eagle Airlines Inc.
13      VX                Virgin America

In [16]:
df = pd.merge(df, airlines, on='AIRLINE')

In [18]:
len(airports)

322

In [19]:
airports.head(2)
# ORIGIN_AIRPORT, DESTINATION_AIRPORT map to the IATA_CODE 

IATA_CODE                              AIRPORT       CITY STATE COUNTRY  \
0       ABE  Lehigh Valley International Airport  Allentown    PA     USA   
1       ABI             Abilene Regional Airport    Abilene    TX     USA   

   LATITUDE  LONGITUDE  
0  40.65236   -75.4404  
1  32.41132   -99.6819

In [20]:
airports.drop('COUNTRY', axis=1, inplace=True)
airports.drop('AIRPORT', axis=1, inplace=True)
airports

IATA_CODE                            CITY STATE  LATITUDE  LONGITUDE
0         ABE                       Allentown    PA  40.65236  -75.44040
1         ABI                         Abilene    TX  32.41132  -99.68190
2         ABQ                     Albuquerque    NM  35.04022 -106.60919
3         ABR                        Aberdeen    SD  45.44906  -98.42183
4         ABY                          Albany    GA  31.53552  -84.19447
..        ...                             ...   ...       ...        ...
317       WRG                        Wrangell    AK  56.48433 -132.36982
318       WYS                West Yellowstone    MT  44.68840 -111.11764
319       XNA  Fayetteville/Springdale/Rogers    AR  36.28187  -94.30681
320       YAK                         Yakutat    AK  59.50336 -139.66023
321       YUM                            Yuma    AZ  32.65658 -114.60597

[322 rows x 5 columns]

In [21]:
origin_airport = airports.copy()
destination_airport = airports.copy()

In [22]:
origin_airport = origin_airport.rename(columns={'IATA_CODE': 'ORIGIN_AIRPORT', 'CITY': 'ORIGIN_CITY', 'STATE': 'ORIGIN_STATE', 'LATITUDE': 'ORIGIN_LATITUDE', 'LONGITUDE': 'ORIGIN_LONGITUDE'})

In [23]:
origin_airport

ORIGIN_AIRPORT                     ORIGIN_CITY ORIGIN_STATE  \
0              ABE                       Allentown           PA   
1              ABI                         Abilene           TX   
2              ABQ                     Albuquerque           NM   
3              ABR                        Aberdeen           SD   
4              ABY                          Albany           GA   
..             ...                             ...          ...   
317            WRG                        Wrangell           AK   
318            WYS                West Yellowstone           MT   
319            XNA  Fayetteville/Springdale/Rogers           AR   
320            YAK                         Yakutat           AK   
321            YUM                            Yuma           AZ   

     ORIGIN_LATITUDE  ORIGIN_LONGITUDE  
0           40.65236         -75.44040  
1           32.41132         -99.68190  
2           35.04022        -106.60919  
3           45.44906         -98.42183  
4           31.53552         -84.19447  
..               ...               ...  
317         56.48433        -132.36982  
318         44.68840        -111.11764  
319         36.28187         -94.30681  
320         59.50336        -139.66023  
321         32.65658        -114.60597  

[322 rows x 5 columns]

In [24]:
destination_airport

IATA_CODE                            CITY STATE  LATITUDE  LONGITUDE
0         ABE                       Allentown    PA  40.65236  -75.44040
1         ABI                         Abilene    TX  32.41132  -99.68190
2         ABQ                     Albuquerque    NM  35.04022 -106.60919
3         ABR                        Aberdeen    SD  45.44906  -98.42183
4         ABY                          Albany    GA  31.53552  -84.19447
..        ...                             ...   ...       ...        ...
317       WRG                        Wrangell    AK  56.48433 -132.36982
318       WYS                West Yellowstone    MT  44.68840 -111.11764
319       XNA  Fayetteville/Springdale/Rogers    AR  36.28187  -94.30681
320       YAK                         Yakutat    AK  59.50336 -139.66023
321       YUM                            Yuma    AZ  32.65658 -114.60597

[322 rows x 5 columns]

In [25]:
destination_airport = destination_airport.rename(columns={'IATA_CODE': 'DESTINATION_AIRPORT', 'CITY': 'DESTINATION_CITY', 'STATE': 'DESTINATION_STATE', 'LATITUDE': 'DESTINATION_LATITUDE', 'LONGITUDE': 'DESTINATION_LONGITUDE'})

In [26]:
destination_airport

DESTINATION_AIRPORT                DESTINATION_CITY DESTINATION_STATE  \
0                   ABE                       Allentown                PA   
1                   ABI                         Abilene                TX   
2                   ABQ                     Albuquerque                NM   
3                   ABR                        Aberdeen                SD   
4                   ABY                          Albany                GA   
..                  ...                             ...               ...   
317                 WRG                        Wrangell                AK   
318                 WYS                West Yellowstone                MT   
319                 XNA  Fayetteville/Springdale/Rogers                AR   
320                 YAK                         Yakutat                AK   
321                 YUM                            Yuma                AZ   

     DESTINATION_LATITUDE  DESTINATION_LONGITUDE  
0                40.65236              -75.44040  
1                32.41132              -99.68190  
2                35.04022             -106.60919  
3                45.44906              -98.42183  
4                31.53552              -84.19447  
..                    ...                    ...  
317              56.48433             -132.36982  
318              44.68840             -111.11764  
319              36.28187              -94.30681  
320              59.50336             -139.66023  
321              32.65658             -114.60597  

[322 rows x 5 columns]

In [27]:
df = pd.merge(df, origin_airport, on='ORIGIN_AIRPORT')

In [28]:
df.head(2)

YEAR  MONTH  DAY  DAY_OF_WEEK AIRLINE  FLIGHT_NUMBER TAIL_NUMBER  \
0  2015      1    1            4      AS             98      N407AS   
1  2015      1    1            4      AS            108      N309AS   

  ORIGIN_AIRPORT DESTINATION_AIRPORT  SCHEDULED_DEPARTURE  ...  \
0            ANC                 SEA                    5  ...   
1            ANC                 SEA                   45  ...   

   AIR_SYSTEM_DELAY  SECURITY_DELAY  AIRLINE_DELAY  LATE_AIRCRAFT_DELAY  \
0               NaN             NaN            NaN                  NaN   
1               NaN             NaN            NaN                  NaN   

   WEATHER_DELAY     airline_full_name  ORIGIN_CITY  ORIGIN_STATE  \
0            NaN  Alaska Airlines Inc.    Anchorage            AK   
1            NaN  Alaska Airlines Inc.    Anchorage            AK   

   ORIGIN_LATITUDE  ORIGIN_LONGITUDE  
0         61.17432        -149.99619  
1         61.17432        -149.99619  

[2 rows x 36 columns]

In [29]:
df = pd.merge(df, destination_airport, on='DESTINATION_AIRPORT')
df

YEAR  MONTH  DAY  DAY_OF_WEEK AIRLINE  FLIGHT_NUMBER TAIL_NUMBER  \
0        2015      1    1            4      AS             98      N407AS   
1        2015      1    1            4      AS            108      N309AS   
2        2015      1    1            4      AS            136      N431AS   
3        2015      1    1            4      AS            134      N464AS   
4        2015      1    1            4      AS            114      N303AS   
...       ...    ...  ...          ...     ...            ...         ...   
5332909  2015     12   29            2      EV           2734      N884AS   
5332910  2015     12   30            3      EV           2730      N880AS   
5332911  2015     12   30            3      EV           2734      N907EV   
5332912  2015     12   31            4      EV           2730      N907EV   
5332913  2015     12   31            4      EV           2734      N907EV   

        ORIGIN_AIRPORT DESTINATION_AIRPORT  SCHEDULED_DEPARTURE  ...  \
0                  ANC                 SEA                    5  ...   
1                  ANC                 SEA                   45  ...   
2                  ANC                 SEA                  135  ...   
3                  ANC                 SEA                  155  ...   
4                  ANC                 SEA                  220  ...   
...                ...                 ...                  ...  ...   
5332909            MEI                 PIB                 2046  ...   
5332910            MEI                 PIB                 1435  ...   
5332911            MEI                 PIB                 2046  ...   
5332912            MEI                 PIB                 1435  ...   
5332913            MEI                 PIB                 2046  ...   

         WEATHER_DELAY            airline_full_name  ORIGIN_CITY  \
0                  NaN         Alaska Airlines Inc.    Anchorage   
1                  NaN         Alaska Airlines Inc.    Anchorage   
2                  NaN         Alaska Airlines Inc.    Anchorage   
3                  NaN         Alaska Airlines Inc.    Anchorage   
4                  NaN         Alaska Airlines Inc.    Anchorage   
...                ...                          ...          ...   
5332909            NaN  Atlantic Southeast Airlines     Meridian   
5332910            0.0  Atlantic Southeast Airlines     Meridian   
5332911            NaN  Atlantic Southeast Airlines     Meridian   
5332912            NaN  Atlantic Southeast Airlines     Meridian   
5332913            NaN  Atlantic Southeast Airlines     Meridian   

         ORIGIN_STATE  ORIGIN_LATITUDE  ORIGIN_LONGITUDE    DESTINATION_CITY  \
0                  AK         61.17432        -149.99619             Seattle   
1                  AK         61.17432        -149.99619             Seattle   
2                  AK         61.17432        -149.99619             Seattle   
3                  AK         61.17432        -149.99619             Seattle   
4                  AK         61.17432        -149.99619             Seattle   
...               ...              ...               ...                 ...   
5332909            MS         32.33313         -88.75121  Hattiesburg-Laurel   
5332910            MS         32.33313         -88.75121  Hattiesburg-Laurel   
5332911            MS         32.33313         -88.75121  Hattiesburg-Laurel   
5332912            MS         32.33313         -88.75121  Hattiesburg-Laurel   
5332913            MS         32.33313         -88.75121  Hattiesburg-Laurel   

         DESTINATION_STATE  DESTINATION_LATITUDE  DESTINATION_LONGITUDE  
0                       WA              47.44898             -122.30931  
1                       WA              47.44898             -122.30931  
2                       WA              47.44898             -122.30931  
3                       WA              47.44898             -122.30931  
4                       WA              47.44898             -122.30931

In [30]:
df.to_csv('cleaned_filghts.csv')

### State Level and Airline Level Calculations

 * As the dataset is very large it makes sense to take a subset of the data to work with.
 * Split for a single state (largest - CA & TX have 600k flights per year = around 12% of dataset each)
 * Split for a single state (medium size @ the 50,000 flights in 2015 level)
 * Split by Airline

In [39]:
state_totals = df.groupby(['ORIGIN_STATE']).ORIGIN_STATE.agg('count')
state_totals.head()

ORIGIN_STATE
AK     33300
AL     26482
AR     22654
AS       107
AZ    165469
Name: ORIGIN_STATE, dtype: int64

In [42]:
state_totals

ORIGIN_STATE
AK     33300
AL     26482
AR     22654
AS       107
AZ    165469
CA    647911
CO    215451
CT     18799
DE       100
FL    415586
GA    360496
GU       334
HI     91181
IA     17382
ID     16915
IL    381644
IN     39107
KS     10966
KY     36880
LA     65185
MA    109290
MD     86079
ME      5047
MI    137033
MN    118701
MO     93838
MS     13109
MT     14848
NC    146906
ND     16530
NE     20023
NH      6292
NJ    108174
NM     22241
NV    146723
NY    246235
OH     74514
OK     31445
OR     55620
PA     99995
PR     26748
RI     11058
SC     28341
SD      9267
TN     75443
TX    631124
UT     99972
VA    138577
VI      5287
VT      2978
WA    123548
WI     51376
WV      2385
WY      8218
Name: ORIGIN_STATE, dtype: int64

In [40]:
airline_totals = df.groupby(['AIRLINE']).AIRLINE.agg('count')
airline_totals.head()

AIRLINE
AA    648694
AS    158054
B6    245135
DL    800329
EV    526249
Name: AIRLINE, dtype: int64

### Let's select Texas as a large state => 631,124 flights in 2015
### And Wisconsin as a 50k pa state to test the workflow!

In [43]:
wisconsin = df[df['ORIGIN_STATE'] == 'WI']
wisconsin = df[df['ORIGIN_STATE'] == 'WI']
wisconsin.to_csv('wisconsin.csv')
florida = df[df['ORIGIN_STATE'] == 'FL']
florida.to_csv('florida.csv')